### ISSUE TITLE
Question about starting sensor values

### ISSUE THREAD

**Issue by TD22057** [ajschmidt8/adext#13]
Thanks for writing this and the HASS alarmdecoder component.  I was having a lot of problems getting my HASS setup working until I upgraded to the lastest dev revision.  I had a few zones where the sensor updates would seem to be getting lost.  I upgraded to the lastest dev version in order to debug that problem - but everything appears to be working ok in that version.

In the process of debugging this, I had to add some logging prints to the alarmdecoder component inside HASS which means I ended up restarting a lot.  Right now, it appears that any zones that aren't toggled have a state "unknown" until they do toggle.  I have ~20 different zones (wireless doors and windows) which I use to track the state of doors and windows and those sensors are basically unusable unless every single one of them get's tripped after every HASS restart.

I realize this isn't a bug and isn't actually a problem on this repo but this seemed like the easiest way to contact you to ask the question:

What do you think about building something that "fixes" this (as much as it can be fixed)?  None of these ways are perfect since zones could be bypassed, etc - but for most cases of restarting HASS, this would work fine.  I'd propose that this mode of operation would be a user option on the component so that if someone would rather have "unknown" vs a possibly wrong state, they can do that.

I've thought of a few possible ways to do this:

1) Use the HASS history component and load the last state change for each entity and set the sensor state to that value.  It might be possible to check the last HASS exit time and only do this if the current time - exit time is less than some value.  i.e. only assume the value is right if HASS hasn't been offline more than x minutes or something like that.

2) Use the panel message.  Unless bypass mode is active, those messages tell us the zones that are faulted or that everything is ok.  That information in alarmdecoder is only reported when it changes but it could be used to set the initial sensor values.

Thoughts?
Ted

**Comment by TD22057** [comment]
3. Change sensors to inherit from RestoreEntity and if a flag is set, call the function to restore their state on startup.  I believe this is the "correct" HASS way of doing this (vs option 1) above.

**Comment by ajschmidt8** [comment]
Hi Ted. Appreciate you reaching out.

Zone tracking in AlarmDecoder currently leaves a lot to be desired. Options 1 and 3 that you've suggested don't seem like they'd be accepted by the Home Assistant folks because it would likely introduce some confusing lines of code that only exist to address a limitation in the AlarmDecoder product itself. And I think implementing that functionality in a third-party package like `adext` would just add some unnecessary complexity.

Option 2 seems feasible. I think the lines of code would look something like:


```py
initial_sensor_state="unknown"
if ALARM_PANEL_STATUS == "READY":
  initial_sensor_state="closed"
```

I'm not sure this is something that I have much time/interest in working on, but if you'd like to open a PR in Home Assistant, I'd be happy to take a look and test it.

**Comment by TD22057** [comment]
Thanks.  I've been really trying to find a way to generally save/load the state for any entity and am not having any luck.  It seems like the internal mechanism for saving states just isn't set up in a way that makes restoring them easy (or maybe possible at all).  I think I'm probably going to drop this and write an MQTT wrapper around the AD code so that I can at least store the sensor on/off messages in the MQTT broker (retain=T) which should restore the sensors states.

**Comment by TD22057** [comment]
Just an FYI but I'm having problems with sensors missing updates.  My wireless sensors routinely don' t have the correct state.  I ran just HA+AD with the latest version of both (latest HA master branch) and it worked fine.  When I use my regular HA config which has more stuff, then things get out of sync.  There is no logging in the AD code anywhere and none in the AD component so I can't really tell where the message disappears.  If I had to guess, I'd say it's in the AD code.  I don't think the way it handles the network comm is very good and uses threads to do the work which might not interact well with HA's async system.  So I'm guessing that AD isn't reading enough and misses information in my case.  That firmed up my desire to write an AD->MQTT bridge (and write a better back end device that does the network coding better) to handle this.  I don't have any way to make a test case for this (other than adding lots of logging to AD+component) but just thought I'd let you know in case anyone else has the same problem.
